Azure Policy extension and assign an Azure Policy for Azure Arc-enabled Kubernetes (#1258)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_k8s/day2/multi_distributions/azure_policy/_index.md

    ![Screenshot showing Azure Portal with Azure Arc-enabled Kubernetes resource extensions](./01.png)

- Edit the environment variables in [the script](https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_k8s_jumpstart/multi_distributions/azure_policy/azure_policy.sh) to match your environment parameters.

  - _`subscriptionId`_ - Your Azure subscription ID
  - _`appId`_ - Your Azure service principal name
  - _`password`_ - Your Azure service principal password
  - _`tenantId`_ - Your Azure tenant ID
  - _`resourceGroup`_ - Azure resource group name
  - _`arcClusterName`_ - Azure Arc Cluster Name
  - _`k8sExtensionName`_ - Azure Policy extension name, should be azurepolicy

    ![Screenshot parameter examples](./02.png)

- After editing the variables, to run the script, navigate to the [script folder](https://github.com/microsoft/azure_arc/tree/main/azure_arc_k8s_jumpstart/multi_distributions/azure_policy) and run the command

  ```shell  
  . ./azure_policy.sh
  ```

    > **NOTE: The extra dot is due to the shell script having an _export_ function and needs to have the vars exported in the same shell session as the rest of the commands.**

   The script will:

>   - Login to your Azure subscription using the SPN credentials

Reviewer: ```suggestion
  - Login to your Azure subscription using the service principal credentials
```

  - Add or Update your local _connectedk8s_ and _k8s-extension_ Azure CLI extensions

>   - Create Azure Policy cluster extension instance

Reviewer: ```suggestion
  - Create the Azure Policy cluster extension instance
```

- Verify under the extensions tab of the Azure Arc-enabled Kubernetes cluster that the Azure Policy cluster extension is correctly installed.

  ![Screenshot showing Azure Portal with Azure Arc-enabled Kubernetes resource extensions](./03.png)

- You can also verify the pods by running the commands below:
  
  - The azure-policy pods are installed in the kube-system namespace:

    ```bash
    kubectl get pods -n kube-system --kubeconfig <cluster-name>.kubeconfig | grep azure-policy
    ```

    ![Screenshot extension pods on cluster](./04.png)

  - The gatekeeper pods are installed in the gatekeeper-system namespace:

    ```bash
    kubectl get pods -n gatekeeper-system --kubeconfig <cluster-name>.kubeconfig
    ```

    ![Screenshot extension pods on cluster](./05.png)

- Verify under the Azure Policy service in the Assignments tab that the Azure Policy Definition was assigned in your resource group:

  ![Screenshot showing Azure Portal Azure Policy service](./06.png)

## Test Azure Policy

> - The Azure Policy we have assigned works as a [LimitRange](https://kubernetes.io/docs/concepts/policy/limit-range/), but we do not specify any namespace, so the Azure Policy will be applied on all namespaces. The limits specified in the Azure Policy are cpuLimit=200m and memoryLimit=1Gi, so to test it we need to create a pod with higher limits:

Reviewer: Better to show steps on how to apply this deployment to the cluster

  ```yaml
  apiVersion: v1
  kind: Pod
  metadata:
    name: nginx
  spec:
    containers:
    - name: nginx
      image: nginx
      resources:
        requests:
          memory: "2Gi"
          cpu: "2"
        limits:
          memory: "3Gi"
          cpu: "16"
  ```

- When we create the above pod, we get the following output:

  ![Screenshot Azure Policy test](./07.png)

> - If there were pods that exceeded the limit before applying the Azure Policy, they will not be removed, but we can see which ones do not comply in the Azure Policy service in the Compliance tab (select the non-compliance event):

Reviewer: ```suggestion
- If there were pods that exceeded the limit before applying the Azure Policy, they will not be removed, but we can see which ones do not comply in the Azure Policy service in the Compliance tab, select the non-compliance event
```

  ![Screenshot Azure Policy test](./08.png)

- Click on Details to find out which pod is not compliant:

  ![Screenshot Azure Policy test](./09.png)

> - This is the non-compliance pod:

Reviewer: ```suggestion
- This is the non-compliant pod:
```
